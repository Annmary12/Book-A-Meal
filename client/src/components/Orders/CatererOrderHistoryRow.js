import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import formatDateString from '../../utils/formatDateString.js';
import localizeNum from '../../utils/localizeNum.js';
import Cell from '../ResponsiveTable/Cell.js';
import Row from '../ResponsiveTable/Row.js';
import ConfirmMarkAsDeliveredModal from './ConfirmMarkAsDeliveredModal.js';
import styles from './order.scss';

/**
 * @class CatererOrderHistoryRow
 */
class CatererOrderHistoryRow extends Component {
  static propTypes = {
    handleDeliver: PropTypes.func.isRequired,
    order: PropTypes.shape({
      id: PropTypes.string,
      quantity: PropTypes.number,
      price: PropTypes.number,
      status: PropTypes.string,
      deliveryAddress: PropTypes.string,
      phoneNumber: PropTypes.string,
      meal: PropTypes.object
    }).isRequired
  }

  state = {
    deliverModalOpen: false
  }

  openDeliverModal = () => {
    this.setState({ deliverModalOpen: true });
  }

  closeDeliverModal = () => {
    this.setState({ deliverModalOpen: false });
  }

  handleDeliver = (orderId) => {
    this.props.handleDeliver(orderId);
    this.closeDeliverModal();
  }


  colorStatus = (status) => {
    let color;
    switch (status) {
      case 'pending':
        color = '#f16800';
        break;
      case 'delivered':
        color = 'green';
        break;
      case 'cancelled':
        color = 'red';
        break;
      default:
        color = 'yellow';
    }

    return (<span style={{ color }}>{status}</span>);
  }

  /**
   * @returns {JSX} React JSX
   */
  render() {
    const { order } = this.props;
    return (
      <Fragment>
        <Row key={order.id}>
          <Cell title="date">
            {formatDateString(order.createdAt)}
          </Cell>
          <Cell title="Meal">
            {order.meal.title}
          </Cell>
          <Cell title="Quantity">
            {order.quantity}
          </Cell>
          <Cell title="Price">
            &#8358;{localizeNum(order.price)}
          </Cell>
          <Cell title="Delivery address">
            <p className={styles.deliveryAddress}>{order.deliveryAddress}</p>
          </Cell>
          <Cell title="Status">
            {this.colorStatus(order.status)}
          </Cell>
          <Cell>
            {
              order.status === 'pending' &&
              <Fragment>
                <button
                  className={styles.deliverBtn}
                  onClick={this.openDeliverModal}
                >
                  Mark as delivered
                </button>
              </Fragment>
            }
          </Cell>
        </Row>
        <ConfirmMarkAsDeliveredModal
          isOpen={this.state.deliverModalOpen}
          handleClose={this.closeDeliverModal}
          deliverOrder={this.handleDeliver}
          orderId={order.id}
        />
      </Fragment>
    );
  }
}

export default CatererOrderHistoryRow;
