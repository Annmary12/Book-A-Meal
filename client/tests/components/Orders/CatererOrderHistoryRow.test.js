import React from 'react';
import CatererOrderHistoryRow from '../../../src/components/Orders/CatererOrderHistoryRow';

const props = {
  handleDeliver: jest.fn(),
  order:  {
    "id": "ecd2344b-9c49-4c78-8623-f7f3334e2f4e",
    "quantity": 1,
    "price": 500,
    "status": "cancelled",
    "deliveryAddress": "24 irepodun street, bodija, ibadan",
    "phoneNumber": "08131928452",
    "createdAt": "2018-09-10T09:05:53.650Z",
    "updatedAt": "2018-09-10T09:06:06.292Z",
    "meal": {
        "id": "dd72bc9b-beb0-49c6-be7b-20106e8aa3d0",
        "title": "Beef with rice",
        "description": "Fried rice and beef",
        "price": 500,
        "img": "https://qph.ec.quoracdn.net/main-qimg-114f40fc7ad4f623ac45ebb33e2c1f8d",
        "caterer": {
            "id": "e20ac257-86cc-4a6f-a619-0249a201c475",
            "fullName": "Otse Obande",
            "username": "otseobande",
            "email": "mealbooker@gmail.com",
            "role": "caterer"
        }
    }
  }
}
describe('CatererOrderHistoryRow component', () => {
  it('should match snapshot if status is not pending', () => {
    const wrapper = shallow(
      <CatererOrderHistoryRow {...props} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if status is pending', () => {
    const updatedProps = {
      ...props,
      order: {
        ...props.order,
        status: 'pending'
      }
    };

    const wrapper = shallow(
      <CatererOrderHistoryRow {...updatedProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if status is delivered', () => {
    const updatedProps = {
      ...props,
      order: {
        ...props.order,
        status: 'delivered'
      }
    };

    const wrapper = shallow(
      <CatererOrderHistoryRow {...updatedProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if status is cancelled', () => {
    const updatedProps = {
      ...props,
      order: {
        ...props.order,
        status: 'cancelled'
      }
    };

    const wrapper = shallow(
      <CatererOrderHistoryRow {...updatedProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });


  it('should match snapshot if status is anything else', () => {
    const updatedProps = {
      ...props,
      order: {
        ...props.order,
        status: 'anything else'
      }
    };

    const wrapper = shallow(
      <CatererOrderHistoryRow {...updatedProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('openDeliverModal method', () => {
    const wrapper = shallow(
      <CatererOrderHistoryRow {...props} />
    );
    it('should set deliverModalOpen state to true', () => {
      wrapper.instance().openDeliverModal();
      const { deliverModalOpen } = wrapper.instance().state;

      expect(deliverModalOpen).toBeTruthy();
    });
  });
  describe('closeDeliverModal method', () => {
    const wrapper = shallow(
      <CatererOrderHistoryRow {...props} />
    );
    it('should set deliverModalOpen state to false', () => {
      wrapper.instance().closeDeliverModal();
      const { deliverModalOpen } = wrapper.instance().state;

      expect(deliverModalOpen).toBeFalsy();
    });
  });

  describe('handleDeliver method', () => {
    const wrapper = shallow(
      <CatererOrderHistoryRow {...props} />
    );

    it('should call handleDeliver prop function', () => {
      const orderId = 'tesd8c';
      wrapper.instance().handleDeliver(orderId);

      expect(props.handleDeliver).toBeCalledWith(orderId);
    });
  });

});