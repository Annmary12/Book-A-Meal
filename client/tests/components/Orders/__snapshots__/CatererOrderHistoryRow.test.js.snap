// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatererOrderHistoryRow component should match snapshot if status is anything else 1`] = `
<Fragment>
  <Row
    header={false}
    key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  >
    <Cell
      title="date"
    >
      9th Sep, 2018
    </Cell>
    <Cell
      title="Meal"
    >
      Beef with rice
    </Cell>
    <Cell
      title="Quantity"
    >
      1
    </Cell>
    <Cell
      title="Price"
    >
      ₦
      500.00
    </Cell>
    <Cell
      title="Delivery address"
    >
      <p
        className="deliveryAddress"
      >
        24 irepodun street, bodija, ibadan
      </p>
    </Cell>
    <Cell
      title="Status"
    >
      <span
        style={
          Object {
            "color": "yellow",
          }
        }
      >
        anything else
      </span>
    </Cell>
    <Cell
      title=""
    />
  </Row>
  <ConfirmMarkAsDeliveredModal
    deliverOrder={[Function]}
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
</Fragment>
`;

exports[`CatererOrderHistoryRow component should match snapshot if status is cancelled 1`] = `
<Fragment>
  <Row
    header={false}
    key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  >
    <Cell
      title="date"
    >
      9th Sep, 2018
    </Cell>
    <Cell
      title="Meal"
    >
      Beef with rice
    </Cell>
    <Cell
      title="Quantity"
    >
      1
    </Cell>
    <Cell
      title="Price"
    >
      ₦
      500.00
    </Cell>
    <Cell
      title="Delivery address"
    >
      <p
        className="deliveryAddress"
      >
        24 irepodun street, bodija, ibadan
      </p>
    </Cell>
    <Cell
      title="Status"
    >
      <span
        style={
          Object {
            "color": "red",
          }
        }
      >
        cancelled
      </span>
    </Cell>
    <Cell
      title=""
    />
  </Row>
  <ConfirmMarkAsDeliveredModal
    deliverOrder={[Function]}
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
</Fragment>
`;

exports[`CatererOrderHistoryRow component should match snapshot if status is delivered 1`] = `
<Fragment>
  <Row
    header={false}
    key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  >
    <Cell
      title="date"
    >
      9th Sep, 2018
    </Cell>
    <Cell
      title="Meal"
    >
      Beef with rice
    </Cell>
    <Cell
      title="Quantity"
    >
      1
    </Cell>
    <Cell
      title="Price"
    >
      ₦
      500.00
    </Cell>
    <Cell
      title="Delivery address"
    >
      <p
        className="deliveryAddress"
      >
        24 irepodun street, bodija, ibadan
      </p>
    </Cell>
    <Cell
      title="Status"
    >
      <span
        style={
          Object {
            "color": "green",
          }
        }
      >
        delivered
      </span>
    </Cell>
    <Cell
      title=""
    />
  </Row>
  <ConfirmMarkAsDeliveredModal
    deliverOrder={[Function]}
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
</Fragment>
`;

exports[`CatererOrderHistoryRow component should match snapshot if status is not pending 1`] = `
<Fragment>
  <Row
    header={false}
    key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  >
    <Cell
      title="date"
    >
      9th Sep, 2018
    </Cell>
    <Cell
      title="Meal"
    >
      Beef with rice
    </Cell>
    <Cell
      title="Quantity"
    >
      1
    </Cell>
    <Cell
      title="Price"
    >
      ₦
      500.00
    </Cell>
    <Cell
      title="Delivery address"
    >
      <p
        className="deliveryAddress"
      >
        24 irepodun street, bodija, ibadan
      </p>
    </Cell>
    <Cell
      title="Status"
    >
      <span
        style={
          Object {
            "color": "red",
          }
        }
      >
        cancelled
      </span>
    </Cell>
    <Cell
      title=""
    />
  </Row>
  <ConfirmMarkAsDeliveredModal
    deliverOrder={[Function]}
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
</Fragment>
`;

exports[`CatererOrderHistoryRow component should match snapshot if status is pending 1`] = `
<Fragment>
  <Row
    header={false}
    key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  >
    <Cell
      title="date"
    >
      9th Sep, 2018
    </Cell>
    <Cell
      title="Meal"
    >
      Beef with rice
    </Cell>
    <Cell
      title="Quantity"
    >
      1
    </Cell>
    <Cell
      title="Price"
    >
      ₦
      500.00
    </Cell>
    <Cell
      title="Delivery address"
    >
      <p
        className="deliveryAddress"
      >
        24 irepodun street, bodija, ibadan
      </p>
    </Cell>
    <Cell
      title="Status"
    >
      <span
        style={
          Object {
            "color": "#f16800",
          }
        }
      >
        pending
      </span>
    </Cell>
    <Cell
      title=""
    >
      <button
        className="deliverBtn"
        onClick={[Function]}
      >
        Mark as delivered
      </button>
    </Cell>
  </Row>
  <ConfirmMarkAsDeliveredModal
    deliverOrder={[Function]}
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
</Fragment>
`;
